File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Micronesia.m
//
//  UIImage+Micronesia.m
//  CountriesFlags
//

#import "Countries/UIImage+Micronesia.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

@implementation UIImage (Micronesia)

+ (UIImage * _Nullable)flagMicronesiaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:117.0f/255.0f
                                         green:178.0f/255.0f
                                          blue:221.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    CGPoint starCenter = CGPointMake(size.width / 2.0f, size.height / 5.0f);
    CGFloat starRadius = size.height / 5.0f / 2.0f;
    CGPoint rotatePoint = CGPointMake(size.width / 2.0f, size.height / 2.0f);
    for (CGFloat angle = 0.0f; angle <= 270.0f; angle += 90.0f) {
        UIBezierPath* path = [UIBezierPath starAtPoint:starCenter
                                                radius:starRadius
                                                  type:FlagStarTypeFivePointed];
        [path rotateToAngle:angle center:rotatePoint];
        [path fill];
    }
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Micronesia.swift
//
//  UIImage+Micronesia.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Federated_States_of_Micronesia "wikipedia"
     
     Generates the flag of the Federated States of Micronesia.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Federated States of Micronesia.
     */
    internal static func flagMicronesia(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 117.0 / 255.0,
                                green: 178.0 / 255.0,
                                blue: 221.0 / 255.0,
                                alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        let starCenter = CGPoint(x: size.width / 2.0, y: size.height / 5.0)
        let starRadius = size.height / 5.0 / 2.0
        let rotatePoint = CGPoint(x: size.width / 2.0, y: size.height / 2.0)
        for angle in stride(from: 0.0, through: 270.0, by: 90.0) {
            let path = UIBezierPath.star(at: starCenter, radius: starRadius)
            path.rotate(to: angle, around: rotatePoint)
            path.fill()
        }

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
